gms/gossiper.cc

```diff
@@ -1660,16 +1661,16 @@ future<> gossiper::handle_major_state_change(inet_address ep, const endpoint_sta
         co_return;
     }
 
-    if (eps_old) {
-        // the node restarted: it is up to the subscriber to take whatever action is necessary
-        co_await _subscribers.for_each([ep, eps_old, pid] (shared_ptr<i_endpoint_state_change_subscriber> subscriber) {
-            return subscriber->on_restart(ep, *eps_old, pid);
-        });
-    }
-
     auto& ep_state = _endpoint_state_map.at(ep);
     if (!is_dead_state(ep_state)) {
         mark_alive(ep, ep_state);
+
+        if (eps_old) {
```

Comment:
How is this code movement related to the rest of the commit?

Reply:
It can actually be refactored out.
It's wrong to notify `on_restart` if the new state `is_dead_state`.

Reply:
Why?

And what if the node went through multiple state transitions: it first restarted in non-dead state, then went to a dead state? Is it possible that we missed the non-dead state moment, and only observed the later moment in dead state, but it is still necessary to call `on_restart` due to that intermediate state that we failed to observe? (But other nodes could have)

Reply:
Generally I'd like to avoid changes like this when there is no known bug that is being fixed, while they might have some unexpected consequences later.

Reply:
fair enough.  will drop and make a note to investigate this as a follow up .
the whole on_restart handling in storage_service:;on_restart seems fishy...